substratevm/src/com.oracle.svm.hosted/src/com/oracle/svm/hosted/phases/NativeImageInlineDuringParsingPlugin.java

```diff
@@ -0,0 +1,545 @@
+                throw debug.handle(ex);
+            }
+            /*
+             * Whatever else happens during the analysis is non-fatal because we can just not inline
+             * that invocation.
+             */
+            return InvocationResult.ANALYSIS_TOO_COMPLICATED;
+        }
+    }
+
+    /**
+     * We collect information during graph construction to filter non-trivial methods.
+     */
+    static class MethodNodeTracking extends NodeEventListener {
+
+        @Override
+        public void nodeAdded(Node node) {
+            if (node instanceof ConstantNode) {
+                /* An unlimited amount of constants is allowed. */
+                /* Nothing to do, an unlimited amount of constants is allowed. We like constants. */
+            } else if (node instanceof ParameterNode) {
+                /* Nothing to do, an unlimited amount of parameters is allowed. */
+                /* Nothing to do. */
+            } else if (node instanceof ReturnNode) {
+                /*
+                 * Nothing to do, returning a value is fine. We don't allow control flow so there
+                 * can never be more than one return.
+                 */
+            } else if (node instanceof LoadFieldNode) {
+                /* Nothing to do, it's ok to read a static or instance field. */
+            } else if (node instanceof FrameState) {
+                if (((FrameState) node).bci == 0) {
+                    /* Nothing to do, it's ok to have frame state for the start node. */
+                } else {
+                    throw new TrivialMethodDetectorBailoutException("Only frame state for the start node is allowed: " + node);
+                }
+            } else if (node instanceof NewArrayNode || node instanceof NewInstanceNode) {
+                /*
+                 * We go further in the analysis to check if any callees can be inline into this
+                 * method. If there was non-constant array or object, method won't be inlined
+                 * because of elimination for existing store field nodes.
+                 */
+            } else {
+                throw new TrivialMethodDetectorBailoutException("Node not allowed: " + node);
+
+            }
+        }
+    }
+
+    /**
+     * Inline trivial invokes (children).
+     */
+    class TrivialChildrenInline implements InlineInvokePlugin {
+
+        @Override
+        public InlineInfo shouldInlineInvoke(GraphBuilderContext b, ResolvedJavaMethod callee, ValueNode[] args) {
+            if (NativeImageInlineDuringParsingPlugin.inliningBeforeAnalysisNotAllowed(b, callee, b.getMethod())) {
+                throw new TrivialMethodDetectorBailoutException("Can't inline: " + callee);
+            }
+
+            CallSite callSite = new CallSite(b.getCallingContext(), NativeImageInlineDuringParsingPlugin.toAnalysisMethod(callee));
+            InvocationResult inline = analyzeMethod(callSite, (AnalysisMethod) callee, args);
```

Comment:
If we call here `analyzeMethod`, we can create infinite recursion. I don't think we should make a new graph in this case, but rather reuse only the analysis.

Reply:
I added fix for this.